### ISSUE TITLE
add an element to the path - need help

### ISSUE THREAD

**Issue by orefalo** [lmorg/murex#680]
**Describe the query:**

I am having the hardest time doing something as simple as adding an element to the `$PATH`

```shell
# bash
$ export PATH="$JAVA_HOME/bin:$PATH"
# fish
$ fish_add_path --path $JAVA_HOME/bin
```
What would be the equivalent code in murex?

Thank you

**Comment by orefalo** [comment]
nevermind

**Comment by lmorg** [comment]
Murex actually has (you guessed it) an undocumented feature for working with PATH and similar. It's the `paths` data type (not to be confused with `path`.

eg
```
tmp » %[/bin, /usr/bin, "$JAVA_HOME/bin"] -> format paths
/bin:/usr/bin:/opt/java/bin
```
(note there's currently a bug with the `%[]` array builder and unquoted strings with variables in, which is why the `$JAVA_HOME` example is quoted but others are not. But I'm hoping to fix that one soon too)

You can use it to split `$PATH` too:
```
» $PATH -> :paths: format json
[
    "/bin",
    "/usr/bin",
    "/opt/java/bin"
]
```
and append too
```
» $PATH -> :paths: append /sbin
/bin:/usr/bin:/opt/java/bin:/sbin
```

basically it turns `$PATH` and similar into arrays that can be queried using all the same array manipulation tools in Murex.

This is very experimental though. Which is why you have to cast it to `paths` first. At some point I'll have `$PATHS`, `$LD_LIBRARY_PATH` and others default to `paths`. Possibly in version 5.0 🤷

**Comment by lmorg** [comment]
I should add that one reason it's considered experimental is because I'm not sold on the name `paths`. It's too similar to `path`. Maybe `dirs` would be better? Or maybe I should have named `path` differently (eg `fs`)?

**Comment by orefalo** [comment]
you have many hidden gems in murex.

I have fixed the above with `$ENV.PATH="$JAVA_HOME/bin:$PATH"`

**Comment by orefalo** [comment]
As far as the name for this hidden functions - :paths: is right on spot IMHO. unless it can be used in other (hidden) contexts. ;-)

on this topic, I am getting a bit confused with all the <keywork>: actions.

ie:

`a:`
`ja:`
`murex-packages:`

and now this `:paths:` which seems to be a different pattern.

**Comment by orefalo** [comment]
thinking loud. you could have

:paths: append /path
:paths: prepend /path
:paths: strip /path
:paths: move /path idx

**Comment by lmorg** [comment]
There's also a `prepend` command: https://dev.murex.rocks/commands/prepend.html

I agree the `append` / `prepend` method is a little klunky. I was thinking of having it supported as an expression. For example overloading `+=`:
```
array += "new item"
```
(that would work for append, less so for prepend)

...but i was worried that operator overloading would lead to things becoming less readable. eg is the following really better than the convention:
```
# with overloading:
$ENV.PATH += "$JAVA_HOME/bin"

# conventional:
$ENV.PATH="$PATH:$JAVA_HOME/bin:"
```

---

Going back to the immediate problem at hand. Once `paths` is fully integrated, it would be possible to write your code like this:
```
ENV.PATH = %[ $JAVA_HOME/bin, @ENV.PATH ]
```

**Comment by orefalo** [comment]
For me, all the above will do. I really like the `ENV.PATH = %[ $JAVA_HOME/bin, @ENV.PATH ]` syntax.

Because there is another issue we are not discussing here, the delimiter is ':' on UNIX, and ';' on M$

**Comment by lmorg** [comment]
> on this topic, I am getting a bit confused with all the : actions

yeah, i haven't documented that well either.

Colon is an optional token that follows the command. The idea was I felt
```
out: hello world
```
was more readable than
```
out hello world
```
(ie just glancing at it would tell you what the command was.

This was in the early days of the Murex shell and I'm really not sure I like it any more but had to keep it in there for compatibility reasons.

Then I thought about having some syntactical sugar for `cast`. I end up doing a lot of casting. eg
```
$PATH -> cast paths -> append $JAVA_HOME/bin -> export PATH
```
so was looking for a way to shorten that. That's where `:type:` comes in. The reason for that was because a lot of other characters like `[]` and `<>` were already taken for other features.

> Because there is another issue we are not discussing here, the delimiter is ':' on UNIX, and ';' on M$

That's already handled:
- https://github.com/lmorg/murex/blob/master/builtins/types/paths/separator_windows.go
- https://github.com/lmorg/murex/blob/master/builtins/types/paths/separator_unix.go

There isn't an override though. So if you're running WSL, cygwin, etc on Windows (ie a Linux build on Windows) then it would default to the UNIX separator. I should probably add this as a `config` option but with the OS determining the default...

**Comment by orefalo** [comment]
The challenge with the `:` notation is that it introduces another classification of commands - making it harder for users to find them (beyond the unusual notation).

Regarding `:paths:`, I can live with the name, maybe not the double `::`.
Still, I am challenging the need to support the functionality as a built-in and not a module. IMHO, the larger the code base, the harder to maintain.  In particular, if you implement something as simple as `ENV.PATH = %[ $JAVA_HOME/bin, @ENV.PATH ]`

Last recommendation - I would not worry too much about backward compatibility. ie. For v5.0 I would shave off all the items marked as deprecated in the previous release.

**Comment by lmorg** [comment]
> The challenge with the `:` notation is that it introduces another classification of commands - making it harder for users to find them (beyond the unusual notation).

It is exactly the same command, just an optional piece of punctuation. tbh I've been considering phasing that out of the documentation.

```
out hello world
```
and 
```
out: hello world
```
are exactly the same thing

> Regarding `:paths:`, I can live with the name, maybe not the double `::`.

How do you mean "double `::`"? as in
```
command -> :type command -> ...
```
rather than
```
command -> :type: command -> ...
```
?

> Still, I am challenging the need to support the functionality as a built-in and not a module. IMHO, the larger the code base, the harder to maintain.  In particular, if you implement something as simple as `ENV.PATH = %[ $JAVA_HOME/bin, @ENV.PATH ]`

I'm not sure I understand your point here either, sorry.

With regards to `ENV.PATH = %[ $JAVA_HOME/bin, @ENV.PATH ]`, there's nothing to maintain. Murex is built around having reusable code. That same line works identically if `$VARIABLE` was a JSON array, CSV file, etc. What does the heavy lifting is the `paths` data type. The code for everything else is unchanged.

**Comment by lmorg** [comment]
That's fine. The documentation sucks. It's impressive you've learned as much as you have :)

**Comment by orefalo** [comment]
I disagree, you wrote a LOT

I will help make it better once the new site is up.

**Comment by orefalo** [comment]
Would you accept a PR to remove all the `:` notation from the examples?
As a beginner, it's been confusing from the start.

ie. https://dev.murex.rocks/commands/f.html#description

**Comment by lmorg** [comment]
yes please!

Pull the version of the code in the `develop` branch and make sure you only edit files named `_doc.yaml` as they're used to generate the markdown files eg https://github.com/lmorg/murex/blob/master/builtins/core/arraytools/2darray_doc.yaml

You'll find the builtins will have their documentation mixed into their source directory. That was intentional because it was easier for me to then keep the docs updated with the source. But it actually doesn't matter where the files are located because it's the `DocumentID` and `CategoryID`s at the top of each YAML document that defines where they get published too.

The actual code within the YAML file is markdown though. eg
```
- DocumentID: 2darray
  Title: >-
    `2darray` 
  CategoryID: commands
  Summary: >-
    # this is written in markdown
  Description: |-
    # this is written in markdown
  Usage: |-
    # this is written in markdown
  Examples: |-
    # this is written in markdown
  Detail: |-
    # this is written in markdown
  Synonyms:
  Related:
  - prepend
  - append
  - map
  - a
  - ja
  - jsplit
  - count
  - index
  - range
  - mtac
  - msort
  - types/json
```
The advantage of this template is it means I can change titles, command names, etc and it gets automatically updated across every document.

eg when I needed to rename the `index.md` document, I could just change the output filename but every document that linked to it was then automatically updated.

It's basically a very hacked together static site "CMS".

---

So edit the `_doc.yaml` files in `develop` and create a PR to merged back into `develop`. Once merged you'll see your changes in `https://dev.murex.rocks`

**Comment by lmorg** [comment]
Good feedback. Thank you

**Comment by orefalo** [comment]
Thanks for merging - I missed a few - will submit a another one.

and then I see that some of the pages are generated from *docgen.go

how can I test locally?

**Comment by lmorg** [comment]
`*docgen.go` files are source code files that have been generated from the CMS. It's how Murex embeds the docs for `help` / `murex-docs` / `[F1]`. You wont need to edit those as they're automatically updated from `docgen`. Plus they're going to change soon because Go now supports embedding files.

---

`*.inc.md` are Markdown include files. It's worth checking those files

---

You can test by running `docgen -config gen/docgen.yaml -warning` in the project root. You'll need to compile `docgen` first though, eg https://github.com/lmorg/murex/blob/develop/.github/workflows/deploy-docs-dev.yaml#L32

**Comment by lmorg** [comment]
> the content menu talks about 'commands', the sidebar talks about 'builtins' - just pick one name and stick to it, it's confusing for beginners

Which do you think would be better? I'm inclined to lean towards 'commands' because that would then allow me to put in anything that isn't technically a builtin. It would also make it cleared when moving stuff like `>>` and `[` out of that section because they technically are builtins even though they appear, superficially at least, like they're not commands.

What's your thought?

**Comment by orefalo** [comment]
Maybe we can also add a section for 'deprecated' as well -

I have a bunch of  ':' changes on my branch, but I am having issues with the latest commit you did. will fix later tonight. 

I will be less active today, sorry

**Comment by lmorg** [comment]
That's fine. No rush. I'm not going to be about much today either :)

**Comment by orefalo** [comment]
I was thinking documentation - 

I think you also need a page to 'itemize', undocumented items

ie.
* :casting notation:
* pre-prompt-func and post-prompt-func in config
* command:
* ...


I can help!

**Comment by orefalo** [comment]
closing as resolved - will update rosetta stone
